README.md
log4c

This is a fork of the version that is located at: http://log4c.sourceforge.net/
This is to honor the license for that version, as I have made changes
to support a project that I am developing. There seems to be no way
to submit changes back to the Source Forge site and that project appears to be
inactive. Others may find my changes useful. 



                                             		Kevin L. Esteb



NEWS
In this release there are now appenders for stderr and stdout.
Previously you would have to use this syntax to output on stderr:

	<category name="root" priority="trace" appender="stderr" />
	<appender name="stderr" type="stream" layout="dated"/>

Where the "stream" type would use the appenders name to open for 
output. Now you can use this syntax:

	<category name="root" priority="trace" appender="testing" />
	<appender name="testing" type="stderr" layout="dated"/>

Which makes the configuration file more compatiable with other
implementations.

1.3.0 2019-11-08:

In this release there is a new static file appender. This appender opens 
the file, writes the log message and then closes the file. This allows 
other processes to monitor the log file. You can now add a cutom parser for
an appender. This allows you to extend the syntax used for an appender.

Summary of the changes:

* added a static file appender.
* added the ability to load a custom parser for an appender.
* removed the basic_r, dated_r and dated_local_r appenders. 

1.2.4   2013-10-03: 

In this release there are new layouts logging messages with local time. There
is also various maintenance work and improved security in public API headers.
Windows MinGW port has been thoroughly tested and updated.

* new date_local and date_local_r appenders
* MinGW updates and important bugfixes:
  * timeval vs FILETIME structures mismatch
  * cleanups
  * support for MinGW >= 4.8
  * add -no-undefied linker flag to create shared dll
  * reentrant layouts on all platforms now
  * replacement functions for buggy snprintf()/vsnprintf()
  * tests working out-of-the-box
  * proper output of long long int
  * generate also .def file
  * providing binaries for the release
* tests working out-of-the-box on FreeBSD
* fixed several warnings
* public API functions with format strings are marked by GNU C format attribute
* updated documentation

1.2.3	2013-04-06:

This release contains important bugfix in syslog appender. There are also
updates for better integration with recent Linux distributions:

* use UTF-8 encoding in AUTHORS file
* update rpm packaging
* update FSF address in license and headers
* manpage for log4c-config
* pkgconfig file
* fixed memory leaks in log4c_ini()/log4c_fini()
* use more recent autotools

1.2.2	2012-10-05:

This is mainly bugfix and portability release, addressing collected bugreports
and patches:

* removed warning in m4 macro
* added log4c_appender_syslog_set_facility()
* rollingfile appender crash fixes
* test updates
* load only first log4crc file
* fixed crash on messages longer than 1024 bytes
* added one missing C++ declaration
* fixed reopening of the stream appender
* proper copy message over end of the file in mmap appender
* portability improvements:
  * support for XCode 3 and OSX 10.5
  * support build with MinGW
  * build tests in FreeBSD
  * try to use C89 compiler (but va_copy or __va_copy still needed)
* various small other fixes

1.2.1	2007-11-08:

* allow use of KB,MB,GB when specifying sizes in config file
	Francisco Camenforte Torres
* memory leak fixes to sd_list, rolling file appender, etc.
	Francisco Camenforte Torres
* hpux syntax fixes
	Francisco Camenforte Torres
* sytax checks during build for c++ compilation

1.2.0	2007-06-21:

- new log4crc hot reread feature. to enable use the --enable-reread
  configure option (Joel Schaubert)
- new API for a logging message with user provided location_info
  (Joel Schaubert)
- added the former XML parser based on libexpat. to disable and
  use the home made parser use the --without-expat configure
  option (Robert Byrne)
- bug fixes

1.1.0	2005-12-16:

- added Windows support from Robert Byrne <Robert.Byrne@Sun.COM>
- added new API log4c_category_list() to retreive a list of the current
  categories being used in log4c
- added an example library contributed by joel schaubert
  <foxyjoel@gmail.com>
- added support for threadsafe loggind and thread safe
  layouts. original idea from Jean-Pierre Joannin
  <jean-pierre.joannin@hp.com>
- added a new API log4c_load() to load/reload a configuration file.
- added new appender stream2 from Robert Byrne <Robert.Byrne@Sun.COM>
- added native Solaris cc support
- added support for suse-8
- log4c got included in gentoo
- added macosx support from Olger Warnier
- plenty of minor fixes

1.0.12	2004-08-18:

This is a release fixing minor bugs concerning native AIX compiler, gcc 3.4.

1.0.11	2003-09-13:

expat was removed and replaced with a simple XML parser.

1.0.10	2002-11-22
1.0.9	2002-11-21
1.0.8	2002-09-23
1.0.7	2002-09-23
1.0.6	2002-05-02
1.0.5	2002-05-02
1.0.4	2002-04-10
1.0.3	2002-03-11
1.0.2	2002-03-11


src/log4c/rc.c
static const char version[] = "$Id: rc.c,v 1.13 2009/05/04 12:30:45 legoater Exp $";

/*
 * rc.c
 *
 * Copyright 2019, Kevin L. Esteb (kevin@kesteb.us). All rights reserved.
 * Copyright 2001-2003, Meiosys (www.meiosys.com). All rights reserved.
 *
 * See the COPYING file for the terms of usage and distribution.
 */

#ifdef HAVE_CONFIG_H
#include "config.h"
#endif

#include <log4c/rc.h>
#include <log4c/category.h>
#include <log4c/appender.h>
#include <log4c/layout.h>
#include <log4c/appender_type_rollingfile.h>
#include <log4c/rollingpolicy.h>
#include <log4c/rollingpolicy_type_sizewin.h>
#include <sd/error.h>
#include <sd/domnode.h>
#include <sd/malloc.h>
#include <sd/sd_xplatform.h>
#include <sd/factory.h>
#include <stdlib.h>
#include <string.h>

 
static log4c_rc_t __log4c_rc = { { 0, 0, 0, 0 } };

log4c_rc_t* const log4c_rc = &__log4c_rc;

/******************************************************************************/
static long parse_byte_size (const char *astring)
{
    /* Parse size in bytes depending on the suffix.   Valid suffixes are KB, MB and GB */
    size_t sz = strlen (astring);
    long res = strtol(astring, (char **) NULL, 10);

    if (res <= 0)
	return 0;

    if (astring[ sz - 1 ] == 'B') {
	switch (astring[ sz - 2 ]) {
	    case 'K':
		res *= 1024;
		break;
	    case 'M':
		res *= 1024 * 1024;
		break;
	    case 'G':
		res *= 1024 * 1024 * 1024;
		break;
	    default:
		sd_debug("Wrong suffix parsing size in bytes for string %s, ignoring suffix", 
			astring);
	}
    }
    sd_debug("Parsed size parameter %s to value %ld",astring, res);
    return (res);
}

/******************************************************************************/
static int config_load(log4c_rc_t* this, sd_domnode_t* anode)
{
    sd_list_iter_t* i = NULL;

    for (i = sd_list_begin(anode->children); i != sd_list_end(anode->children); 
	 i = sd_list_iter_next(i)) 
    {
	sd_domnode_t* node = i->data;

	if (!strcmp(node->name, "nocleanup")) {
	    this->config.nocleanup = atoi(node->value);
	    if (this->config.nocleanup)
		sd_debug("deactivating log4c cleanup");		
	}

	if (!strcmp(node->name, "bufsize")) {	    
	    this->config.bufsize = parse_byte_size(node->value);

	    if (this->config.bufsize)
		sd_debug("using fixed buffer size of %d bytes", 
			 this->config.bufsize);
	    else
		sd_debug("using dynamic allocated buffer");
	}

	if (!strcmp(node->name, "debug")) {
	    sd_domnode_t* level = sd_domnode_attrs_get(node, "level");

	    if (level) {
		this->config.debug = atoi(level->value);
		sd_debug("activating log4c debugging. level = %d", this->config.debug);
	    }
	}
        if (!strcmp(node->name, "reread")) {
            this->config.reread = atoi(node->value);
            sd_debug("log4crc reread is %d",this->config.reread);
            if (0 == this->config.reread)
                sd_debug("deactivating log4crc reread");
        }
    }
    
    return 0;
}

/******************************************************************************/
static int category_load(log4c_rc_t* this, sd_domnode_t* anode)
{
    sd_domnode_t*     name     = sd_domnode_attrs_get(anode, "name");
    sd_domnode_t*     priority = sd_domnode_attrs_get(anode, "priority");
    sd_domnode_t*     additivity = sd_domnode_attrs_get(anode, "additivity");
    sd_domnode_t*     appender = sd_domnode_attrs_get(anode, "appender");
    log4c_category_t* cat      = NULL;
    
    if (!name) {
	sd_error("attribute \"name\" is missing");
	return -1;
    }

    cat = log4c_category_get(name->value);
    
    if (priority)
	log4c_category_set_priority(
	    cat, log4c_priority_to_int(priority->value));
    
    if (additivity) {
	if (!strcasecmp(additivity->value, "false")) {
	    log4c_category_set_additivity(cat, 0);
	} else if (!strcasecmp(additivity->value, "true")) {
	    log4c_category_set_additivity(cat, 1);
	} else {
	    sd_error("additivity value is invalid : %s", additivity->value);
	}
    }

    if (appender)
	log4c_category_set_appender(
	    cat, log4c_appender_get(appender->value));

    return 0;
}

/******************************************************************************/
static int appender_load(log4c_rc_t* this, sd_domnode_t* anode)
{
    sd_domnode_t*     name   = sd_domnode_attrs_get(anode, "name");
    sd_domnode_t*     type   = sd_domnode_attrs_get(anode, "type");
    sd_domnode_t*     layout = sd_domnode_attrs_get(anode, "layout");
    log4c_appender_t* app    = NULL;

    if (!name) {
        sd_error("attribute \"name\" is missing");
        return -1;
    }
    
    sd_debug("appender_load[name='%s'", 
      (name->value ? name->value :"(not set)"));
    
    app = log4c_appender_get(name->value);

    if (type) {

        sd_debug("appender type is '%s'",
                 (type->value ? type->value: "(not set)"));

        log4c_appender_set_type(app, log4c_appender_type_get(type->value));
        log4c_appender_parse(app, anode);

    }

    if (layout) {
        log4c_appender_set_layout(app, log4c_layout_get(layout->value));
    }

    sd_debug("]");

    return 0;

}

/******************************************************************************/
static int layout_load(log4c_rc_t* this, sd_domnode_t* anode)
{
    sd_domnode_t*   name   = sd_domnode_attrs_get(anode, "name");
    sd_domnode_t*   type   = sd_domnode_attrs_get(anode, "type");
    log4c_layout_t* layout = NULL;
    
    if (!name) {
	sd_error("attribute \"name\" is missing");
	return -1;
    }

    layout = log4c_layout_get(name->value);

    if (type)
	log4c_layout_set_type(layout, log4c_layout_type_get(type->value));

    return 0;
}

#ifdef WITH_ROLLINGFILE
/******************************************************************************/
static int rollingpolicy_load(log4c_rc_t* this, sd_domnode_t* anode)
{
    sd_domnode_t*   name   = sd_domnode_attrs_get(anode, "name");
    sd_domnode_t*   type   = sd_domnode_attrs_get(anode, "type");
    log4c_rollingpolicy_t* rpolicyp = NULL;
    long a_maxsize;
    
    sd_debug("rollingpolicy_load[");
    if (!name) {
	sd_error("attribute \"name\" is missing");
	return -1;
    }

    rpolicyp = log4c_rollingpolicy_get(name->value);    

    if (type){
	log4c_rollingpolicy_set_type(rpolicyp,
             log4c_rollingpolicy_type_get(type->value));
    
      if (!strcasecmp(type->value, "sizewin")){
        sd_domnode_t*   maxsize   = sd_domnode_attrs_get(anode, "maxsize");
        sd_domnode_t*   maxnum  = sd_domnode_attrs_get(anode, "maxnum");
        rollingpolicy_sizewin_udata_t *sizewin_udatap = NULL;
        
        sd_debug("type='sizewin', maxsize='%s', maxnum='%s', "
                  "rpolicyname='%s'",
           (maxsize && maxsize->value ? maxsize->value :"(not set)"),
           (maxnum && maxnum->value ? maxnum->value :"(not set)"),
           (name && name->value ? name->value :"(not set)"));
        /*
         * Get a new sizewin policy type and configure it.
         * Then attach it to the policy object.
         * Check to see if this policy already has a
         sw udata object.  If so, leave as is except update
         the params
        */
        if ( !(sizewin_udatap = log4c_rollingpolicy_get_udata(rpolicyp))){ 
          sd_debug("creating new sizewin udata for this policy");
          sizewin_udatap = sizewin_make_udata();
          log4c_rollingpolicy_set_udata(rpolicyp,sizewin_udatap);   
	  a_maxsize = parse_byte_size(maxsize->value);
	  if (a_maxsize)
            sizewin_udata_set_file_maxsize(sizewin_udatap, a_maxsize);
	  else{
	    sd_debug("When parsing %s a size of 0 was returned. Default size %d will be used",
	      maxsize->value, ROLLINGPOLICY_SIZE_DEFAULT_MAX_FILE_SIZE);
	    sizewin_udata_set_file_maxsize(sizewin_udatap, ROLLINGPOLICY_SIZE_DEFAULT_MAX_FILE_SIZE); 
	  }

        sizewin_udata_set_max_num_files(sizewin_udatap, atoi(maxnum->value));
        }else{
          sd_debug("policy already has a sizewin udata--just updating params");
        sizewin_udata_set_file_maxsize(sizewin_udatap, parse_byte_size(maxsize->value));
        sizewin_udata_set_max_num_files(sizewin_udatap, atoi(maxnum->value));
         /* allow the policy to initialize itself */
        log4c_rollingpolicy_init(rpolicyp, 
            log4c_rollingpolicy_get_rfudata(rpolicyp));
        }        
       
      }
    
    }
    sd_debug("]");

    return 0;
}
#endif

/******************************************************************************/
extern int log4c_rc_load(log4c_rc_t* this, const char* a_filename)
{    
    sd_list_iter_t* i = NULL;
    sd_domnode_t*   node = NULL;        
    sd_domnode_t*   root_node = NULL;

    sd_debug("parsing file '%s'\n", a_filename);
    
    if (!this)
	return -1;

    root_node = sd_domnode_new(NULL, NULL);
    
    if (sd_domnode_load(root_node, a_filename) == -1) {
	sd_domnode_delete(root_node);
	return -1;
    }

    /* Check configuration file root node */
    if (strcmp(root_node->name, "log4c")) {
	sd_error("invalid root name %s", root_node->name);
	sd_domnode_delete(root_node);
	return -1;
    }
    
    /* Check configuration file revision */
    if ( (node = sd_domnode_attrs_get(root_node, "version")) != NULL)
	if (strcmp(VERSION, node->value)) {
	    sd_error("version mismatch: %s != %s", VERSION, node->value);
	    sd_domnode_delete(root_node);
	    return -1;
	}
	    
    /* backward compatibility. */
    if ( (node = sd_domnode_attrs_get(root_node, "cleanup")) != NULL) {
	sd_debug("attribute \"cleanup\" is deprecated");
	this->config.nocleanup = !atoi(node->value);
    }
			
    /* load configuration elements */
    for (i = sd_list_begin(root_node->children);
	 i != sd_list_end(root_node->children); 
	 i = sd_list_iter_next(i)) 
    {
	sd_domnode_t* node = i->data;

	if (!strcmp(node->name, "category")) category_load(this, node);
	if (!strcmp(node->name, "appender")) appender_load(this, node);
#ifdef WITH_ROLLINGFILE
  if (!strcmp(node->name, "rollingpolicy"))rollingpolicy_load(this, node);
#endif
	if (!strcmp(node->name, "layout"))   layout_load(this, node);
	if (!strcmp(node->name, "config"))   config_load(this, node);
    }
	
    sd_domnode_delete(root_node);
    
    return 0;
}

/******************************************************************************/
extern int log4c_load(const char* a_filename)
{
    return log4c_rc_load(&__log4c_rc, a_filename);
}
